# **Score-based linkage and partial agreement**

## Introduction 

In the **Data preparation and deterministic linkage script,** you saw a pipeline that:
* Cleaned the data ready for linkage
* Performed exact matching
* Derived new relaxed variables
* Performed exact matching using matchkeys
* Created and read linkage metrics 

This script will take you through:
* a recap of data preparation
* a recap of exact and rule-based linkage, including the use of perturbed (messy) data
* blocking
* using Levenshtein edit distance and
* running score-based linkage and using Levenshtein edit distance to inform match scores.

You will also go through some of the methods to investigate whether they need adjusting to increase the true positives. 

A few tips to start with:
* As in the **Data preparation and deterministic linkage script,** for this code to work, the directory needs to be the same for all files so everything needs to be saved in the same place. Make sure the data files and the code are all saved in the same folder.
* This script mostly contains functions. Ensure functions are run before they are called otherwise it will cause an error message.
* The code below uses more complex methods compared to the **Data preparation and deterministic linkage script.** Make sure you run through and understand the first script before moving onto this one.



# **Data preparation**

You have two datasets which you want to link **practitioner_large_file** & **practitioner_small_file**:

- **practitioner_large_file** contains the variables `IDA`, `Sex`, `Locality`, `Yearbirth`, `Monthbirth`, `Daybirth`, `Name` and a record ID that is contained in the variable `Bident`. In addition, there is a variable `Aident` that contains the record ID from the small file that it is matched to, i.e. we know the true match status.

- **practitioner_small_file** contains the variables `IDB`, `Sex`, `Locality`, `Yearbirth`, `Monthbirth`, `Daybirth`, `Name` and a record ID that is contained in the variable `Aident`. In addition there is a variable, `Bident`, that contains the record ID from the large file that it is matched to, i.e. we know the true match status.

Some of the variables in the **practitioner_small_file** were perturbed to simulate measurement errors. A perturbed variable is a variable whose value has been intentially altereed, usually by a small amount. Here this is done to simulate errors in real world data. It may also be used to protect record anonimity when publushing results, for example when publishing population statistics or census results for a small town. The perturbed data here are called `Sexpert`, `Yearbirthpert`, `Monthbirthpert`, `Daybirthpert`, `Namepert`. `Locality` was not perturbed and we will use this as a blocking variable.

### Importing packages
As in the **Data preparation and deterministic linkage script,** the below code imports the relevant packages and modifies the display of information so it is easier to vied in jupyter.

In [33]:
# Importing packages, modify settings and widen display
import sys
import os
import pandas as pd
import numpy as np

%pip install rapidfuzz
from rapidfuzz.distance import Levenshtein

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

pd.set_option('display.width', 1000)

Looking in indexes: https://bestj:****@onsart-01/artifactory/api/pypi/yr-python/simpleNote: you may need to restart the kernel to use updated packages.




[notice] A new release of pip is available: 24.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


### Get file path function
This function locates the directory (filepath) on your computer. This directory is used in subsequent functions to upload the files needed to work this notebook. Make sure you have all the relevant files (data and code) saved in the same folder.

In [34]:
# get_file_path function
def get_file_path():
    """
    Return the current working directory.

    This utility function retrieves the absolute path of the directory from which
    the code is being executed. It is typically used alongside functions such as
    `read_data` to establish a consistent working directory for file operations
    within a Jupyter notebook or script.

    Returns
    -------
    str
        The absolute path to the current working directory.
    """
    file_path = os.getcwd()
    return file_path

### Read data function
Below is a function that reads in the data and ensures the data types are correct.  

In [35]:
#Below is a function that reads in the data
def read_data(filepath):
    """ 
    Load the practitioner datasets from the specified directory and return them as DataFrames.

    This function reads two CSV files—`practitioner_large_file.csv` and
    `practitioner_small_file.csv`—from the provided file path. It also applies
    consistent dtype coercion to ensure that key variables (e.g., sex, birthdate
    components, locality, and names) are loaded with the expected types for
    downstream cleaning, matching, and performance evaluation.

    Parameters
    ----------
    filepath : str
        The directory path where the input CSV files are located. The directory
        must contain:
        - ``practitioner_large_file.csv``  
        - ``practitioner_small_file.csv``

    Returns
    -------
    tuple of pandas.DataFrame
        A tuple ``(dfA, dfB)`` where:
        - ``dfA`` is the large practitioner dataset.
        - ``dfB`` is the smaller practitioner dataset.
        Both DataFrames have selected columns coerced into consistent numeric or
        string types to support subsequent processing stages.

    """
    
    pd.set_option('display.width', 1000)
    
    dfA = pd.read_csv(filepath + '\\practitioner_large_file.csv') 
    dfB = pd.read_csv(filepath + '\\practitioner_small_file.csv') 
    
  
    dfA = dfA.astype({"SEX_A": float,
                  "locality_A": float,
                  "yearbirth_A": float,
                  "monthbirth_A": float,
                  "daybirth_A": float,
                  "Name_A": str})
    
    dfB = dfB.astype({"SEX_B": float,
                  "locality_B": float,
                  "yearbirth_B": float,
                  "monthbirth_B": float,
                  "daybirth_B": float,
                  "Name_B": str})
    
    return (dfA, dfB)

Have a look at the data.

In [36]:
# Call the two functions and display the columns headings and first 20 rows as well as counts
filepath = get_file_path()
dfA, dfB = read_data(filepath)
dfA.head()
print(len(dfA))
dfB.head()
print(len(dfB))

,SEX_A,locality_A,yearbirth_A,monthbirth_A,daybirth_A,Name_A,bident_A,aident_A,ID_A
0,1.0,31.0,1958.0,12.0,3.0,geoff,E373,A1,2051588
1,1.0,31.0,1965.0,1.0,30.0,james,E371,NaN,2047429
2,1.0,70.0,1922.0,7.0,28.0,Michael,E348,NaN,1926422
3,1.0,70.0,1946.0,1.0,NaN,terry,E346,NaN,1907224
4,1.0,70.0,1949.0,9.0,6.0,MATTHEW,E332,A2,1818785


1000


,SEX_B,locality_B,yearbirth_B,monthbirth_B,daybirth_B,Name_B,namepert_B,yearbirthpert_B,monthbirthpert_B,daybirthpert_B,sexpert_B,aident_B,bident_B,ID_B
0,1.0,7400.0,1975.0,7.0,6.0,alan,alan,75.0,7,6.0,1.0,A245,E925,5196460
1,1.0,7900.0,1975.0,9.0,30.0,JulianDickin,Julian,1076.0,9,30.0,1.0,A266,E20,97397
2,2.0,962.0,1900.0,6.0,NaN,Naomi,Naomi,1900.0,6,NaN,2.0,A377,E621,3563159
3,2.0,6500.0,1905.0,3.0,NaN,debbie,debbie,1905.0,4,NaN,2.0,A544,E903,5115700
4,1.0,3000.0,1905.0,3.0,NaN,Mike,Mike,1905.0,3,NaN,1.0,A95,E291,1576027


680


### Standardising variables functions - upper case name function 
As in the **Data preparation and deterministic linkage script,** the variables need to be standardised to ensure comparing like with like. Below is a function that makes all name columns uppercase. 

<div class="alert alert-block alert-danger">
Note that namepert_B is a perturbed variable.  
</div>

In [37]:
# Upper case name function
def upper_case_name(dfA, dfB):
    """
    Convert name-related columns in two dataframes to uppercase.

    This function looks for a predefined set of column names in each of the
    provided DataFrames. When a matching column is found, its string values
    are converted to uppercase in place.

    Parameters
    ----------
    dfA : pandas.DataFrame
        The first input DataFrame to update.
    dfB : pandas.DataFrame
        The second input DataFrame to update.

    Returns
    -------
    tuple of pandas.DataFrame
        A tuple containing the updated (dfA, dfB) DataFrames.

    """
    
    columns = ["Name_A", "Name_B", "namepert_B"]
    
    for df in (dfA, dfB):
        for name in columns:
            if name in df.columns:   
                df[name] = df[name].str.upper()
                
    return dfA, dfB

### Call the function
The below code calls the above function.

In [38]:
# Call the function and have a look at the data
dfA, dfB = upper_case_name(dfA, dfB)
dfB. head(20)

,SEX_B,locality_B,yearbirth_B,monthbirth_B,daybirth_B,Name_B,namepert_B,yearbirthpert_B,monthbirthpert_B,daybirthpert_B,sexpert_B,aident_B,bident_B,ID_B
0,1.0,7400.0,1975.0,7.0,6.0,ALAN,ALAN,75.0,7,6.0,1.0,A245,E925,5196460
1,1.0,7900.0,1975.0,9.0,30.0,JULIANDICKIN,JULIAN,1076.0,9,30.0,1.0,A266,E20,97397
2,2.0,962.0,1900.0,6.0,NaN,NAOMI,NAOMI,1900.0,6,NaN,2.0,A377,E621,3563159
3,2.0,6500.0,1905.0,3.0,NaN,DEBBIE,DEBBIE,1905.0,4,NaN,2.0,A544,E903,5115700
4,1.0,3000.0,1905.0,3.0,NaN,MIKE,MIKE,1905.0,3,NaN,1.0,A95,E291,1576027
5,1.0,7800.0,1907.0,3.0,19.0,PETER,DETER,1907.0,3,19.0,1.0,A254,E885,5029982
6,2.0,6200.0,1908.0,12.0,20.0,CHRISTINE,CHRISTINE,1908.0,12,20.0,2.0,A527,E191,945609
7,1.0,8700.0,1908.0,5.0,NaN,PAUL,PAUL,1908.0,5,NaN,1.0,A301,E464,2619977
8,2.0,8300.0,1909.0,11.0,6.0,ANITA,ANITA,1909.0,11,6.0,2.0,A622,E255,1343394
9,1.0,188.0,1909.0,5.0,1.0,JOHN,JOHN,1909.0,5,1.0,1.0,A14,E627,3575957


# **Recap: Exact linkage**

### Exact matching using unperturbed variables
As in the **Data preparation and deterministic linkage script,** you will first perform exact linkage on the two datasets to examine the number of 'easy' matches. dfB is a cut of dfA so there should be no unmatched records in dfB. 

In [39]:
# exact_link_with_key_mapping function
def exact_link_with_key_mapping(left, right, key_map, how, indicator):
    """
    Merge two DataFrames whose join keys have different names.

    Parameters
    ----------
    left, right : pd.DataFrame
    key_map : dict
        Mapping from left key name -> right key name.
        Supports multiple keys, e.g. {'Age': 'age_years', 'DOB': 'birth_date'}
    how : str
        'left', 'right', 'inner', 'outer'
     indicator : bool
         if True, adfd pandas merge indicator column '_merge'.
         
    Returns
    -------
    pd.DataFrame
    """
    left_on = list(key_map.keys())
    right_on = [key_map[k] for k in left_on]

   
    linked = pd.merge(
        left, right,
        how=how,
        left_on=left_on,
        right_on=right_on,
        indicator=indicator
    )

    return linked

### Call the function 
The below code calls the function merge with key mapping, merging dfA amd dfB. 

In [40]:
# Call the function merge_with_key_mapping
key_map = {
    "SEX_A": "SEX_B",
    "locality_A": "locality_B",
    "yearbirth_A": "yearbirth_B",
    "monthbirth_A": "monthbirth_B",
    "daybirth_A": "daybirth_B",
    "Name_A": "Name_B"
}

linked_df1 = exact_link_with_key_mapping(dfA, dfB, key_map, how="inner", indicator=True)
residuals1 = exact_link_with_key_mapping(dfA, dfB, key_map, how="outer", indicator=True)
print('Number of records matched: ',len(linked_df1))


Number of records matched:  680


In [41]:
# Have a look at how many matches are made
residualsA = residuals1[residuals1['_merge'] == 'left_only']
print('Number of unmatched records in dataframeA: ',len(residualsA))

residualsB = residuals1[residuals1['_merge'] == 'right_only']
print('Number of unmatched records in dataframeB: ',len(residualsB))

Number of unmatched records in dataframeA:  320
Number of unmatched records in dataframeB:  0


As expected the are 680 matched records from dfB to dfA as the unperturbed variables have been used. Let's merge the perturbed variables and see how many matches are made. 

### Exact linkage using perturbed variables
Below calls the linkage on dfA and the perturbed variables in dfB. 

In [42]:
key_map = {
    "SEX_A": "sexpert_B",
    "locality_A": "locality_B",
    "yearbirth_A": "yearbirthpert_B",
    "monthbirth_A": "monthbirthpert_B",
    "daybirth_A": "daybirthpert_B",
    "Name_A": "namepert_B"
}

linked_df2 = exact_link_with_key_mapping(dfA, dfB, key_map, how="inner", indicator=True)
residuals2 = exact_link_with_key_mapping(dfA, dfB, key_map, how="outer", indicator=True)

Remember that there are 0 unmatched variables in dfB, how many are there now? 

In [43]:
# Have a look at residuals made
residualsB = residuals2[residuals2['_merge'] == 'right_only']
print('Number of unmatched records in dataframeB: ',len(residualsB))

Number of unmatched records in dataframeB:  175


This time, 175 records from the smaller dataset (dfB) did not link to a record in the larger dataset (dfA) due to errors in the data. Let's have a look at the residuals from dfB to see what types of errors caused names not to be matched. 

**Question: What sort of errors can you see?**

In [44]:
# View names of clean & perturbed dataset B residuals - errors may have prevented exact match
residualsB[['Name_B', 'namepert_B']]

,Name_B,namepert_B
5,MATTHEW,MATTMEN
7,MICAELL,MICAELU
8,NEIL,MEIL
14,JOE,SOF
16,JOHN,SOHN
...,...,...
1170,BOB,BOB
1171,JAMES,JAMES
1172,ELEANOR,BEANOR
1173,JENNIFER,JENNIFER


**Answer:**
* Discrepancy in date perturbed variables (discrepancies and errors in how year of birth is recorded in yearbirthpert_B)
* Two names in the unperturbed variable Name_B 

### Standardising variables functions - parse and clean names function
The below function will parse (split) and clean names to standardise and correct some of the errors found in dfB.

In [45]:
# Parse and clean names function that standardises the name variables to increase linkage
def parse_and_clean_names(df, input_col, out_col1, out_col2):
    """
    Parse a name field into two components and clean title placements.

    This function takes a column containing a two-part name (usually first name and
    surname, but potentially including a prefix such as 'MR' or 'MRS'). It splits
    the string on the first space into two parts, assigns them to the specified
    output columns, and then performs title‑aware corrections:

    - If the first token is a title (e.g. 'MR', 'MRS'), the two parts are swapped
      so that the non-title becomes the first component.
    - If the second token is a title, it is removed and replaced with ``None``.

    The result is a cleaner two-column representation of names, suitable for
    downstream matching and preprocessing steps.

    Parameters
    ----------
    df : pandas.DataFrame
        The DataFrame containing the name column to be split.
    input_col : str
        The name of the column containing the full name string to be split.
    out_col1 : str
        Name of the output column for the first name component.
    out_col2 : str
        Name of the output column for the second name component.

    Returns
    -------
    pandas.DataFrame
        The input DataFrame with two additional columns (``out_col1`` and
        ``out_col2``) containing the cleaned name components.

    """
    df[[out_col1, out_col2]] = df[input_col].str.split(' ', expand=True)
    
    titles = {'MR', 'MRS'}

    df[out_col1], df[out_col2] = np.where(
        df[out_col1].isin(titles),
        [df[out_col2], df[out_col1]],
        [df[out_col1], df[out_col2]]
    )

    df[out_col2] = np.where(
        df[out_col2].isin(titles),
        None,
        df[out_col2]
    )

    return df


In [46]:
#Call the function for relevamt name variables
dfA = parse_and_clean_names(dfA, 'Name_A', 'Name1_A', 'Name2_A')
dfB = parse_and_clean_names(dfB, 'namepert_B', 'Name1pert_B', 'Name2pert_B')

In [47]:
# Have a look at the dataframe A to check the changes made for dfA
dfA[['Name1_A', 'Name2_A']].head()

,Name1_A,Name2_A
0,GEOFF,None
1,JAMES,None
2,MICHAEL,None
3,TERRY,None
4,MATTHEW,None


In [48]:
# Have a look at the changes made for dfB
dfB[['Name1pert_B', 'Name2pert_B']].head()

,Name1pert_B,Name2pert_B
0,ALAN,None
1,JULIAN,None
2,NAOMI,None
3,DEBBIE,None
4,MIKE,None


Try the merge again with improved and standardised name variables to see how many matches are made.

In [49]:
# Run exact_link_with_key_mapping function for matches and residuals (as in the Data preparation and deterministic linkage script))
key_map = {
    "SEX_A": "sexpert_B",
    "locality_A": "locality_B",
    "yearbirth_A": "yearbirthpert_B",
    "monthbirth_A": "monthbirthpert_B",
    "daybirth_A": "daybirthpert_B",
    "Name1_A": "Name1pert_B"
}

linked_df2 = exact_link_with_key_mapping(dfA, dfB, key_map, how="inner", indicator=True)
residuals2 = exact_link_with_key_mapping(dfA, dfB, key_map, how="outer", indicator=True)

How many unmatched records are there? 

In [50]:
# Check residuals from dfB
residualsB = residuals2[residuals2['_merge'] == 'right_only']
print('Number of unmatched records in dataframeB: ',len(residualsB))

Number of unmatched records in dataframeB:  167


There are still 167 unmatched records in `dfB`. Examine the residuals, what kind of errors in the data are still there?

In [51]:
# Have a look at the residuals
residualsB[['Name_B', 'Name1pert_B']]

,Name_B,Name1pert_B
5,MATTHEW,MATTMEN
7,MICAELL,MICAELU
8,NEIL,MEIL
14,JOE,SOF
16,JOHN,SOHN
...,...,...
1162,BOB,BOB
1163,JAMES,JAMES
1164,ELEANOR,BEANOR
1165,JENNIFER,JENNIFER


# **Rule-based linkage**

As in the **Data preparation and deterministic linkage script,** here you can use rule-based linkage to relax name and increase the number of matches. Be aware that this will probably increase the number of false positives. 

You will do this by creating a short_names function as in the **Data preparation and deterministic linkage script.**

In [52]:
# Short_names function
def parse_short_names(df):
    columns = ["Name1_A", "Name1pert_B"]

    for name in columns:
        if name in df.columns:
            df["short" + name] = (
                df[name]
                .where(df[name].notna())     
                .astype("string")            
                .str.slice(0, 3)             
            )

    return df

In [53]:
# Call the parse_short_names function, outline new variables for merge and merge the two dataframes using the two new varaibles
dfA = parse_short_names(dfA)
dfB = parse_short_names(dfB)

key_map = {
    "SEX_A": "sexpert_B",
    "locality_A": "locality_B",
    "yearbirth_A": "yearbirthpert_B",
    "monthbirth_A": "monthbirthpert_B",
    "daybirth_A": "daybirthpert_B",
    "shortName1_A":"shortName1pert_B"
    }

linked_df3 = exact_link_with_key_mapping(dfA, dfB, key_map, how="inner", indicator=True)
print('Number of matched records in dataframe: ',len(linked_df3))

Number of matched records in dataframe:  565


Remember that there were 167 unmatched records in dfB before. How many are there now?

In [54]:
# Run the below code to check residuals 
residuals3 = exact_link_with_key_mapping(dfA, dfB, key_map, how="outer", indicator=True)
residualsB = residuals3[residuals3['_merge'] == 'right_only']
print('Number of unmatched records in dataframeB: ',len(residualsB))

Number of unmatched records in dataframeB:  115


Great! Now you only have 115 unmatched records - BUT are all of the matches correct? 

Check to see if ID_A=ID_B for all the records in the linked dataset. Run the next function to find out.

### Linkage metrics function - evaluate_matches
As in the **Data preparation and deterministic linkage script,** you will need to evaluate the quality of the matches produced in df3 by looking at true positives and false positives. Here, you create a match_status (as in exercie 1) and identify all true positives as 1 and all false positives as 0. 

The below function first ensures the columns input into the function exist in the dataframe.

<div class="alert alert-block alert-warning">
missing = [c for c in (id_a, id_b) if c not in df.columns]<br>
    if missing:<br>
        raise KeyError(f"Missing required column(s): {missing}")<br>
</div>

Then it creates match_status to clearly mark which links are matches and which are not using id_a and id_b as the 'ground truth'.

<div class="alert alert-block alert-warning">
df_out[match_col] = np.where(df_out[id_a] == df_out[id_b], 1, 0)
</div>

Again, note that here you use id_a and id_b as they are common between the two datasets. In reality, you may not have a variable in common between the two datasets or the common variable quality may be too low to use. In this eventuality, have another look at the **Data preparation and deterministic linkage script** which mentions different methods that can be used instead to evaluate the quality of matches made. 

In the code, you then allocate true positives as 1 in the match column and false positives as 0 which it then prints out for ease using the *verbose* function.

<div class="alert alert-block alert-warning">
true_positives = df_out[df_out[match_col] == 1]<br>
    false_positives = df_out[df_out[match_col] == 0]<br>
</div>

The next section of code prints out the number of true positives and false positives. 

<div class="alert alert-block alert-warning">
if verbose:<br>
        print("Number of true positives (matches): ", len(true_positives))<br>
        print("Number of false positives (non-matches): ", len(false_positives))<br>
</div>




In [55]:
def evaluate_matches(df, id_a,id_b, match_col, verbose):
    """
    Add a match status column to a dataframe, print counts, and return true/false positives.

    Parameters
    ----------
    df : pd.DataFrame
        Input dataframe containing the two identifier columns to compare.
    id_a : str, default "ID_A"
        Name of the column representing the A-side ID.
    id_b : str, default "ID_B"
        Name of the column representing the B-side ID.
    match_col : str, default "Match_Status"
        Name of the output column that flags matches (1) and non-matches (0).
    verbose : bool, default True
        If True, prints the counts of true positives and false positives.

    Returns
    -------
    df_out : pd.DataFrame
        The original dataframe with the new `match_col` column added.
    true_positives : pd.DataFrame
        Subset of rows where `match_col == 1` (IDs match).
    false_positives : pd.DataFrame
        Subset of rows where `match_col == 0` (IDs do not match) — kept to match user's terminology.
    """
    missing = [c for c in (id_a, id_b) if c not in df.columns]
    if missing:
        raise KeyError(f"Missing required column(s): {missing}")

    df_out = df.copy()

    df_out[match_col] = np.where(df_out[id_a] == df_out[id_b], 1, 0)

    true_positives = df_out[df_out[match_col] == 1]
    false_positives = df_out[df_out[match_col] == 0]

    if verbose:
        print("Number of true positives (matches): ", len(true_positives))
        print("Number of false positives (non-matches): ", len(false_positives))

    return df_out, true_positives, false_positives

In [56]:
# Call and define the columns to inspect for TPs and FPs
tp_cols = [
    'SEX_A', 'sexpert_B',
    'yearbirth_A', 'yearbirthpert_B',
    'monthbirth_A', 'monthbirthpert_B',
    'daybirth_A', 'daybirthpert_B',
    'Name_A', 'namepert_B'
]

fp_cols = [
    'SEX_A', 'sexpert_B',
    'yearbirth_A', 'yearbirthpert_B',
    'monthbirth_A', 'monthbirthpert_B',
    'daybirth_A', 'daybirthpert_B',
    'Name_A', 'namepert_B'
]

linked_df3_with_status, TPs, FPs = evaluate_matches(
    linked_df3,
    id_a="ID_A",
    id_b="ID_B",
    match_col="Match_Status",
    verbose=True
)

Number of true positives (matches):  561
Number of false positives (non-matches):  4


There are four false positives. Go back to the start of creating a short name variable and change the length of the Short_Name_A and Short_Name_B. 

**Question: What is the best length of string to use to find the most true positives with the least false positives?** 

To answer this question, you will need to run the short_names function and amend the code *.str.slice(0, 3)* to another number.

**Answer:**
* If you decrease the number of letters, there will be more matches but a higher likelihood of a false negative match being made. Increasing the number to 4 will increase the number of true matches and decrease the number of false positives. 

### Rule-based matching - relax date of birth
What will happen if you relax date of birth by letting day be a mismatch?

In [57]:
# Run exact_link_with_key_mapping for matches and residuals
key_map = {
    "SEX_A": "sexpert_B",
    "locality_A": "locality_B",
    "monthbirth_A": "monthbirthpert_B",
    "yearbirth_A": "yearbirthpert_B",
    "Name1_A": "Name1pert_B"
}

linked_df4 = exact_link_with_key_mapping(dfA, dfB, key_map, how="inner", indicator=True)

How many residuals are there now?

In [58]:
# Count residuals
residuals4 = exact_link_with_key_mapping(dfA, dfB, key_map, how="outer", indicator=True)
residualsB = residuals4[residuals4['_merge'] == 'right_only']

print('Number of unmatched records in dataframeB: ',len(residualsB))

Number of unmatched records in dataframeB:  159


**Question: How many of the matches are correct? Run the evaluation function on these matches.**

In [59]:
# Define the columns to inspect for TPs and FPs
tp_cols = [
    'SEX_A', 'sexpert_B',
    'yearbirth_A', 'yearbirthpert_B',
    'monthbirth_A', 'monthbirthpert_B',
    'daybirth_A', 'daybirthpert_B',
    'Name_A', 'namepert_B'
]

fp_cols = [
    'SEX_A', 'sexpert_B',
    'yearbirth_A', 'yearbirthpert_B',
    'monthbirth_A', 'monthbirthpert_B',
    'daybirth_A', 'daybirthpert_B',
    'Name_A', 'namepert_B'
]

linked_df4_with_status, TPs, FPs = evaluate_matches(
    linked_df4,
    id_a="ID_A",
    id_b="ID_B",
    match_col="Match_Status",
    verbose=True
)



Number of true positives (matches):  520
Number of false positives (non-matches):  1


There is a false positive. Change the variables in the merges to see if we should remove a different, or another, date of birth variable.

In [60]:
# Run the code below to see if you need to change the
#date of birth variable to be removed.
FPs = linked_df4_with_status[linked_df4_with_status['Match_Status']==0]
FPs[['SEX_A', 'sexpert_B',
     'yearbirth_A', 'yearbirthpert_B',
     'monthbirth_A', 'monthbirthpert_B',
     'daybirth_A', 'daybirthpert_B',
     'Name_A', 'namepert_B']]

,SEX_A,sexpert_B,yearbirth_A,yearbirthpert_B,monthbirth_A,monthbirthpert_B,daybirth_A,daybirthpert_B,Name_A,namepert_B
240,1.0,1.0,1967.0,1967.0,7.0,7,15.0,14.0,ADAM,ADAM


**Answer:**
* Both year and month of birth are the same whereas day of birth is different. This means that there will be a false positive if day of birth is used for a match. 
* To make sure you only have true positives, change the missing variable in the key map to either month or year instead of day. 

#**Blocking**
Blocking can be used to reduce the search area by only using non missing, identical values on certain variables. These are paired as possible matches, called blocking variables. Remember that picking your blocking variable is crucial so bear in mind to:
* use accurately recorded variables
* use variables with complete coverage
* blocking can be done on more than one variable
* if blocks are too large, the search space is not reduced enough.

Once your blocking variables have been picked, they will divide your data into smaller chunks ready for any probabilistic linkage method you want to run on the data. Remember that cutting down the number of comparisons can risk missing true matches, so a balance is needed.

If you do not have a viable blocking variable, which may be the case for your datasets, the below code outliines one possible way of blocking *without* a specific blocking variable. You will generate all possible pairs using the perturbed variables.

In [61]:
# You can do this by creating two columns containing a single value and then join on that value
dfA['join_col'] = 0
dfB['join_col'] = 0

key_map = {
    "join_col":"join_col"
}

CP1 = exact_link_with_key_mapping(dfA, dfB, key_map, how="inner", indicator=True)
print('There are', len(CP1), 'possible pairs. Of these, 680 (0.1%) are true matches.')

There are 680000 possible pairs. Of these, 680 (0.1%) are true matches.


`dfA` contains 1,000 records and `dfB` contains 680 records.

In [62]:
print('There are', len(CP1), 'possible pairs. Of these, 680 (0.1%) are true matches.')

There are 680000 possible pairs. Of these, 680 (0.1%) are true matches.


### Blocking - with a blocking variable
Now you will use the blocking variable `locality`. This isn't perturbed hence the expectation is that it will retain all 680 true matches (and also find extra, false matches). Bear in mind when using blocking in your dataset that using geographical location may not be suitable if the population you are linking is highly mobile and the datasets collection points are at different times of year. The population may have moved between the two dates and records may not link despite two records being a match.

In [63]:
key_map = {
    "locality_A":"locality_B"
}

CP2 = exact_link_with_key_mapping(dfA, dfB, key_map, how="left", indicator=True)
print('There are', len(CP2), 'possible pairs. Of these, 680 (3.6%) are true matches.')

There are 18908 possible pairs. Of these, 680 (3.6%) are true matches.


# **Score Based Matching**
Now to move onto score-based matching. This should take place after exact and rule-based matching to help you deal with any remaining ambiguous pairs.

As a reminder, the advantages of score-based linkage are:
* Score-based matching can draw out any remaining matches between your datasets after other methods have been exhausted, and account for different types of error.
* You can control the number of false positives and false negatives by changing the score threshold.
* The classification of a match is less subjective, making linkage results more statistically robust.

The disadvantages are:
* Score-based linkage is computationally expensive to run. This means it may not always be feasible in resource limited environments.
* Similarly, score-based linkage, especially when using more complex techniques like Fellegi-Sunter, requires a lot of time and effort to design. Again this means it may not always be feasible in resource or time limited environments (such as in an emergency or crisis).
* To get the most out of score-based linkage, some clerical matching is required. 

The below function computes the score of a matching pair. It does this by:
* Assigning a weight to matching variables based on quality and distinguishability.
* It then multiplies agreement and weights, summing them for a total score.

Remember that missing values do not necessarily mean that they do not match and they would therefore be coded as matching when using this function. This is standard practice in linkage protocols.

First the function defines the default parameter weights and variables. If these need to be changed you can amend them when the function is called.

<div class="alert alert-block alert-warning">
(df: pd.DataFrame,
                        name_w=5, sex_w=1, day_w=2, month_w=2, year_w=3,<br>
                        name_a='Name1_A', name_b='Name1pert_B',<br>
                        sex_a='SEX_A',  sex_b='sexpert_B',<br>
                        day_a='daybirth_A',   day_b='daybirthpert_B',<br>
                        mon_a='monthbirth_A', mon_b='monthbirthpert_B',<br>
                        yr_a='yearbirth_A',   yr_b='yearbirthpert_B',<br>
                        out_col='Score')
</div>

The calculation for a pair wise score is *weight * agreement*. Therefore, this function includes this calculation for each variable and assigns it to a new column within the dataframe.

<div class="alert alert-block alert-warning">
score = (<br>
        ((df[name_a] == df[name_b]) * name_w) +<br>
        ((df[sex_a]  == df[sex_b])  * sex_w)  +<br>
        (((df[day_a] == df[day_b])  | (np.isnan(df[day_a]) & np.isnan(df[day_b])))   * day_w)   +<br>
        (((df[mon_a] == df[mon_b])  | (np.isnan(df[mon_a]) & np.isnan(df[mon_b])))   * month_w) +<br>
        (((df[yr_a]  == df[yr_b])   | (np.isnan(df[yr_a])  & np.isnan(df[yr_b])))    * year_w)<br>
    )<br>
    df[out_col] = score.astype(int)
</div>

In [64]:
# Compute_score_basic function
def compute_score_basic(df: pd.DataFrame,
                        name_w=5, sex_w=1, day_w=2, month_w=2, year_w=3,
                        name_a='Name1_A', name_b='Name1pert_B',
                        sex_a='SEX_A',  sex_b='sexpert_B',
                        day_a='daybirth_A',   day_b='daybirthpert_B',
                        mon_a='monthbirth_A', mon_b='monthbirthpert_B',
                        yr_a='yearbirth_A',   yr_b='yearbirthpert_B',
                        out_col='Score1'):
    """
    Compute a weighted exact‑match similarity score between paired records.

    This function evaluates whether selected variables from dataset A and
    dataset B match, assigns a predefined weight to each matching variable,
    and sums these weights to produce an overall similarity score. For the
    day, month, and year of birth fields, pairs in which both values are NaN
    are treated as matches. All comparisons are exact (no fuzzy matching).

    Parameters
    ----------
    df : pandas.DataFrame
        The DataFrame containing paired comparison variables from datasets A and B.
    name_w, sex_w, day_w, month_w, year_w : int, optional
        Weights applied when the corresponding fields match. Defaults are:
        name=5, sex=1, day=2, month=2, year=3.
    name_a, name_b : str
        Column names for the name variables in datasets A and B.
    sex_a, sex_b : str
        Column names for the sex variables.
    day_a, day_b : str
        Column names for the day‑of‑birth variables.
    mon_a, mon_b : str
        Column names for the month‑of‑birth variables.
    yr_a, yr_b : str
        Column names for the year‑of‑birth variables.
    out_col : str, optional
        The name of the output column to store the final score. Default is "Score".

    Returns
    -------
    pandas.Series
        A Series containing the integer similarity score for each row of `df`.

    """
    score = (
        ((df[name_a] == df[name_b]) * name_w) +
        ((df[sex_a]  == df[sex_b])  * sex_w)  +
        (((df[day_a] == df[day_b])  | (np.isnan(df[day_a]) & np.isnan(df[day_b])))   * day_w)   +
        (((df[mon_a] == df[mon_b])  | (np.isnan(df[mon_a]) & np.isnan(df[mon_b])))   * month_w) +
        (((df[yr_a]  == df[yr_b])   | (np.isnan(df[yr_a])  & np.isnan(df[yr_b])))    * year_w)
    )
    df[out_col] = score.astype(int)
    return df[out_col]


Let's run the above function on the dataset without the blocking variable.

In [65]:
# Call function and compute scores for dataset with no blocking variables
compute_score_basic(CP1, out_col='Score1')
CP1.Score1.value_counts(sort=False)


0         1
1         1
2         0
3         0
4         1
         ..
679995    4
679996    3
679997    3
679998    4
679999    0
Name: Score1, Length: 680000, dtype: int64

Score1
1     285290
0     303336
2      38742
3      41339
5       1758
13       513
6       2422
4       6159
8        342
9         22
7         21
12        14
11        26
10        16
Name: count, dtype: int64

There are 513 exact matches as expected - the same as the number of residuals after `name` cleaning and exact matching. Below you will run the dataset using the blocking variable to compare.

In [66]:
# Call function and compute scores for dataset using locality as a blocking variable
compute_score_basic(CP2, out_col='Score2')
CP2.Score2.value_counts(sort=False)

0        13
1         0
2         1
3         0
4         0
         ..
18903     0
18904     0
18905     1
18906     1
18907     8
Name: Score2, Length: 18908, dtype: int64

Score2
13     513
0     8071
1     7682
2     1050
3     1127
8      107
4      156
7        5
12      14
9        1
5       64
6       80
11      23
10      15
Name: count, dtype: int64

As predicted, when not using a blocking variable you compare many more potential matching pairs, hence the large numbers at 0 and 1 when looking at score1 compared to score2.

### Setting Thresholds
Thresholds are a cut off point that determine whether a pair of records is accepted as a true match. Remember that you would ideally have two score thresholds and three regions. You would do this by
* Examining the top of the sorted list for pairs with the highest scores that do not match.
* When they start appearing frequently fix the upper threshold.
* Coninue to look through the list until matches become moderately rare.
* Then fix the lower threshold.

Here you will be shown how to create one threshold using the scores from CP2 and how to evaluate whether it is effective. 

The below function includes one threshold, the default being anything equal to or above 10. It evaluates the weighted scores using the threshold as a marker of a correct match and evaluates this against the *ground truth* of true positives and false positives using id_a and id_b. 

<div class="alert alert-block alert-warning">
df['Match_Status'] = (df[id_a] == df[id_b]).astype(int)
</div>

It then calculates the predicted matches using the score and threshold to compare results.

<div class="alert alert-block alert-warning">
df['Predicted_Match'] = (df[score_col] >= threshold).astype(int)
</div>

A subset dataframe of true matches using the predicted matches is created for ease later. 

<div class="alert alert-block alert-warning">
matches = df[df['Predicted_Match'] == 1].copy()
</div>

Then labels them as true positives and false positives.

<div class="alert alert-block alert-warning">
tp = matches['Match_Status'].sum()<br>
    fp = len(matches) - tp
</div>

It then prints out the results. 

<div class="alert alert-block alert-warning">
print(f"With threshold {threshold}, there are {len(matches)} predicted matches.")<br>
    print("Number of correct matches:", tp)<br>
    print("Number of false positives:", fp)
</div>


In [67]:
# Evaluate_matches function
def evaluate_matches(
    df,
    score_col,
    id_a='ID_A',
    id_b='ID_B',
    threshold=10
):
    """
    Evaluate linkage matches and write results directly into the dataframe.

    This function adds two new columns:
        - 'Match_Status'     : 1 if ID_A == ID_B else 0
        - 'Predicted_Match'  : 1 if score >= threshold else 0

    It also returns a dictionary summarising:
        - threshold
        - number of predicted matches
        - true positives
        - false positives
        - the matched subset (DataFrame)
    """

    df['Match_Status'] = (df[id_a] == df[id_b]).astype(int)
    df['Predicted_Match'] = (df[score_col] >= threshold).astype(int)
    matches = df[df['Predicted_Match'] == 1].copy()

    tp = matches['Match_Status'].sum()
    fp = len(matches) - tp

    print(f"With threshold {threshold}, there are {len(matches)} predicted matches.")
    print("Number of correct matches:", tp)
    print("Number of false positives:", fp)

    return {
        'threshold': threshold,
        'n_matches': len(matches),
        'true_positives': tp,
        'false_positives': fp,
    }

In [68]:
# Call the threshold function for CP2
evaluate_matches(
    df= CP2,
    score_col='Score2',
    id_a='ID_A',
    id_b='ID_B',
    threshold=10)


With threshold 10, there are 565 predicted matches.
Number of correct matches: 559
Number of false positives: 6


{'threshold': 10,
 'n_matches': 565,
 'true_positives': np.int64(559),
 'false_positives': np.int64(6)}

**Question: What do the different scores tell you?** 
You can alter the weights and threshold until you are happy that you have maximised true positives and minimised false positives.

**Answer:**
Make sure the weights you decide to use make sense. Remember 
* higher weights for variables that are more unique
* for variables that are reliably recorded
* check scores empirically as well as intuitively (what is the score distribution)?

In altering the threshold, remember that you may increase the false positive rate if your threshold is to low. 

### String comparator - Levenshtein edit distance function
String comparators compare different words to measure how similar they are. Particularly if working with administrative data, spelling mistakes are common. Here you will be using Levenshtein edit distance. Remember that it measures distance rather than similarity, measuring the number of deletions, insertions or substitutions to transform one string into another. As you are using the normalized Levenshtein edit distance (which accounts for length of word), tt gives a result between 0 (perfect disagreement) and 1. 

<div class="alert alert-block alert-danger">
Remember that the method of gathering the data is important when linking two datasets. For example, if one dataset has collected data by phone a more appropriate tool may be to use in built phonetic functions to compare two strings for the purpose of linkage as different people will not necessarily spell the same name in the same way. 
</div>

In [69]:
def SLD(s, t):
    """
    Compute the Standardised Levenshtein Distance (SLD) similarity score.

    This function returns a similarity value between 0 and 1 based on the
    normalized Levenshtein similarity measure. A score of 1.0 indicates
    perfect similarity (identical strings), while 0.0 indicates complete
    dissimilarity.

    The implementation uses RapidFuzz's `Levenshtein.normalized_similarity`,
    which provides a fast and efficient computation optimized in C++.

    Parameters
    ----------
    s : str
        The first string to compare.
    t : str
        The second string to compare.

    Returns
    -------
    float
        A normalized similarity score in the range [0, 1].

    """
    return Levenshtein.normalized_similarity(s, t)

In [70]:
# Call the function
string1="Rachel"
string2="Rachael"
print('Standardised Levenshtein Edit Distance [',string1,',',string2,'] = ',SLD(string1, string2))

Standardised Levenshtein Edit Distance [ Rachel , Rachael ] =  0.8571428571428572


### Score Based Matching with Partial Scores - Compute score function
With string comparators you can perform score based matching using partial scores. There are many reasons why this may be suitable: 
* it can handle imperfect and inconsistent data (e.g. with spelling mistakes)
* provides more discrimination than binary matching
* it supports weighted evidence across variables (i.e. it weighs certain variables higher than others as you have done in the scoring. This can be useful when performing linkage and you are aware of certain variables being more reliable than others.)
* you can distinguish links using a threshold based decision rule (i.e. if the threshold is two, this means that the rule will only accept two differences between two strings being compared).

The below function uses the levenshtein distance function previously created to assess how similar two strings are (here these are names). Now that you have replaced the nan values with empty values the only things being compared will be names. In order to calculate the new scores, you will re-calculate the previous scores and multiply them by the results of Levenshtein distance.

As before when you calculated the score without the string_comparator, you wil outline the weights in the function signature. 

The function changes NaN's to empty stings. This is because
* as Levenshtein only expects strings, this prevents comparisons between 'nan' (string) as a word and a name
* allows partial similarity

<div class="alert alert-block alert-warning">
df[name_a] = df[name_a].fillna('')
</div>

It the function then creates the Levenshtein similarity score ready to be multiplied by the weight. 

<div class="alert alert-block alert-warning">
df['_SLD'] = df.apply(lambda row: lev_func(row[name_a], row[name_b]), axis=1)
</div>

The code then only accepts exact matches. 

<div class="alert alert-block alert-warning">
sex_score  = (df[sex_a]  == df[sex_b]).astype(float)<br>
    day_score  = (df[day_a]  == df[day_b]).astype(float)<br>
    mon_score  = (df[mon_a]  == df[mon_b]).astype(float)<br>
    year_score = (df[yr_a]   == df[yr_b]).astype(float)
</div>

It then calculates the new score which includes the new Levenshtein *score * name_weight*.
 

In [71]:
# Computing score with levenshtein results
def compute_score_with_levenshtein(
    df: pd.DataFrame,
    lev_func,
    name_w=5, sex_w=1, day_w=2, month_w=2, year_w=3,
    name_a='Name1_A', name_b='Name1pert_B',
    sex_a='SEX_A',  sex_b='SEX_B',
    day_a='daybirth_A',   day_b='daybirthpert_B',
    mon_a='monthbirth_A', mon_b='monthbirthpert_B',
    yr_a='yearbirth_A',   yr_b='yearbirthpert_B',
    out_col='Score2'
):
    """
    Compute a weighted comparison score between paired records, replicating
    the behaviour of the legacy CP2 matching code.

    This function:
      • Replaces missing name values (NaN) with empty strings before computing
        similarity.
      • Applies a supplied Levenshtein-based string similarity function to all
        name pairs (including empty–empty pairs).
      • Treats day, month, and year of birth as exact‐match comparisons only:
        missing–missing is NOT treated as a match.
      • Applies user-specified weights to each component and returns the
        weighted score.

    Parameters
    ----------
    df : pandas.DataFrame
        DataFrame containing the paired record comparison fields.

    lev_func : callable
        A function accepting two strings and returning a numeric similarity
        (typically in [0, 1]), such as a normalised Levenshtein similarity.

    name_w, sex_w, day_w, month_w, year_w : int or float, optional
        Weights applied to the name, sex, day, month, and year comparison
        components respectively.

    name_a, name_b : str, optional
        Column names for the name fields.

    sex_a, sex_b : str, optional
        Column names for the sex fields.

    day_a, day_b : str, optional
        Column names for the day-of-birth fields.

    mon_a, mon_b : str, optional
        Column names for the month-of-birth fields.

    yr_a, yr_b : str, optional
        Column names for the year-of-birth fields.

    out_col : str, optional
        Name of the output column in which the score will be stored.

    Returns
    -------
    pandas.Series
        A Series containing the computed weighted similarity score for each row.

    """

    df[name_a] = df[name_a].fillna('')
    df[name_b] = df[name_b].fillna('')

    df['_SLD'] = df.apply(lambda row: lev_func(row[name_a], row[name_b]), axis=1)

    sex_score  = (df[sex_a]  == df[sex_b]).astype(float)
    day_score  = (df[day_a]  == df[day_b]).astype(float)
    mon_score  = (df[mon_a]  == df[mon_b]).astype(float)
    year_score = (df[yr_a]   == df[yr_b]).astype(float)

    df[out_col] = (
        df['_SLD'] * name_w +
        sex_score  * sex_w  +
        day_score  * day_w  +
        mon_score  * month_w +
        year_score * year_w
    )

    df.drop(columns=['_SLD'], inplace=True)

In [72]:
# Call the function
compute_score_with_levenshtein(
    df=CP2,
    lev_func=SLD,
    name_w=5, sex_w=1, day_w=2, month_w=2, year_w=3,
    out_col='score_with_SLD'
)

In [73]:
# Evaluate the results
results = evaluate_matches(
    df= CP2,
    score_col='score_with_SLD',
    id_a='ID_A',
    id_b='ID_B',
    threshold=10)

With threshold 10, there are 646 predicted matches.
Number of correct matches: 636
Number of false positives: 10


**Question: Change the `Threshold` to see if you can make the false postive rate smaller. What threshold would you need to reduce the number of false positives to 0?**

You can do this by amending the threshold when you call the evaluate_matches function

**Answer:**
A threshold of 12 reduces the number of false positives to 0.

### Export dataframe
The code below will export the dataframe out of this notebook.

In [74]:
CP2.to_csv('CP2_export.csv', index=False)